024-generalized-resources/proposal.md

```diff
@@ -0,0 +1,297 @@
+
+* Support [trigger resources](../024-trigger-resources/proposal.md).
+
+* Support [spatial resources](../024-spatial-resources/proposal.md).
+
+* Support [notification resources](../024-notification-resources/proposal.md).
+
+## Glossary
+
+* **Config**: an arbitrarily nested JSON object containing user-provided configuration
+  * Examples: `{"uri":"https://github.com/concourse/concourse"}`, `{"interval":"10m"}`
+
+* **Config fragment**: a smaller JSON object intended to be "spliced" into a **config** by assigning each field from the fragment into the config.
+  * Examples: `{"ref":"abcdef"}`, `{"branch":"develop"}`
+
+* **Bits**: a directory containing arbitrary data
+  * Examples: source code, compiled artifacts, etc.
+
+* **Metadata**: structured data associated to a **config fragment** containing information about the fragment that should be surfaced to the user
+  * Examples: `[{"name":"committer","value":"Alex Suraci"}]`
+
+* **Resource type**: an implementation of the interface defined by this proposal, typically provided as a container image. Implements the following actions:
+  * `info`: given a **config**, emit provide the commands to run for the following actions:
+  * `check`: given a **config**, emit **config fragments**
+  * `get`: given a **config**, populate a directory with **bits**
+  * `put`: given a **config** and a directory containing **bits**, create or delete **config fragments**
+  * Examples:
+    * `git-branches` resource type for tracking branches in a repo
+    * `git` resource type for tracking commits in a branch
+    * `github-status` resource type for emitting build status notifications for commits
+    * `time` resource type for doing timed job triggers
+
+* **Resource**: a **resource type** with a user-provided **config**, used together to represent external state.
+
+## Example Resources
+
+```yaml
+type: git-branches
+source:
+  uri: https://github.com/concourse/concourse
+```
+
+```yaml
+type: git
+source:
+  uri: https://github.com/concourse/concourse
+  branch: master
+```
+
+```yaml
+type: github-status
+source:
+  repository: concourse/concourse
+  access_token: abcdef
+```
+
+```yaml
+type: time
+source:
+  interval: 10m
+```
+
+## Interface Types
+
+```go
+// Config represents arbitrary user-specified configuration.
+type Config map[string]interface{}
+
+// ConfigFragment represents additional fields that can be spliced into a Config.
+type ConfigFragment map[string]interface{}
+
+// MetadataField represents a named bit of metadata associated to a ConfigFragment.
+type MetadataField struct {
+  Name  string `json:"name"`
+  Value string `json:"value"`
+}
+
+// TLSConfig captures common configuration for communicating with servers over TLS.
+type TLSConfig struct {
+  // An array of CA certificates to trust.
+  CAs []string `json:"ca_certs,omitempty"`
+
+  // Disable TLS, effectively making communication over TLS insecure.
+  InsecureSkipVerify bool `json:"insecure_skip_verify,omitempty"`
```

Comment:
This wording feels strange but I *think* I've seen this variation the most.

If I were to name this I would have called it `skip_verification`, but I guess the idea is to shove 'insecure' in there somewhere so people know not to use it. I'm not sure about 'skip_verify' vs 'skip_verificication' though - to me the latter feels more natural. :man_shrugging: